File: 프로그래머스/LV2/코딩테스트 연습/탐욕법/구명보트/구명보트.cpp
#include <string>
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

int solution(vector<int> people, int limit) {
    int answer = 0;
    sort(people.begin(), people.end());

    deque<int> q(people.begin(), people.end());

    while (q.size() > 1) {
        if(q.front() + q.back() <= limit) {
            q.pop_front();
            q.pop_back();
        }
        else {
            q.pop_back();
        }
        answer++;
    }

    if(q.size())answer++;

    return answer;
}

File: 프로그래머스/LV2/코딩테스트 연습/탐욕법/구명보트/구명보트.py
from collections import deque

def solution(people, limit):
    ans = 0
    people.sort()
    
    q = deque(people)

    while len(q) > 1:
        if q[0] + q[-1] <= limit:
            q.popleft()
            q.pop()
        else:
            q.pop()
        ans += 1
    
    if q: ans += 1

    return ans
